java_gen/pre-written/src/main/java/org/projectfloodlight/openflow/util/HexString.java

```diff
@@ -1,100 +1,202 @@
-/**
- *    Copyright (c) 2008 The Board of Trustees of The Leland Stanford Junior
- *    University
- *
- *    Licensed under the Apache License, Version 2.0 (the "License"); you may
- *    not use this file except in compliance with the License. You may obtain
- *    a copy of the License at
- *
- *         http://www.apache.org/licenses/LICENSE-2.0
+package org.projectfloodlight.openflow.util;
+
+import io.netty.util.internal.EmptyArrays;
+
+/** Utility method to convert hexadecimal string from/to longs and byte arrays.
  *
- *    Unless required by applicable law or agreed to in writing, software
- *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
- *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
- *    License for the specific language governing permissions and limitations
- *    under the License.
- **/
+ * @author Andreas Wundsam <andreas.wundsam@bigswitch.com>
+ */
+public final class HexString {
 
-package org.projectfloodlight.openflow.util;
+    private HexString() {}
 
-import org.projectfloodlight.openflow.types.U8;
+    /* ================================================================================
+     * Implementation note:
+     * These implementations are optimized for small-O efficiency and thus rather ugly.
+     *
+     * When making changes, make sure *every line* is covered by a unit test.
+     *
+     * Do not use this as a blue print for normal code.
+     * ================================================================================
+     */
+
+    private final static char[] CHARS =
+        { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };
 
-public class HexString {
     /**
      * Convert a string of bytes to a ':' separated hex string
      *
      * @param bytes
      * @return "0f:ca:fe:de:ad:be:ef"
      */
     public static String toHexString(final byte[] bytes) {
-        int i;
-        String ret = "";
-        String tmp;
-        for (i = 0; i < bytes.length; i++) {
-            if (i > 0)
-                ret += ":";
-            tmp = Integer.toHexString(U8.f(bytes[i]));
-            if (tmp.length() == 1)
-                ret += "0";
-            ret += tmp;
+        int lenBytes = bytes.length;
+        if (lenBytes == 0) {
+            return "";
         }
-        return ret;
-    }
+        char arr[] = new char[lenBytes * 2 + (lenBytes -1)];
 
-    public static String toHexString(final long val, final int padTo) {
-        char arr[] = Long.toHexString(val).toCharArray();
-        String ret = "";
-        // prepend the right number of leading zeros
-        int i = 0;
-        for (; i < (padTo * 2 - arr.length); i++) {
-            ret += "0";
-            if ((i % 2) != 0)
-                ret += ":";
+        int charPos = 0;
+        for (int i=0; i < lenBytes; i++) {
+            if (i > 0) {
+              arr[charPos++] = ':';
```

Comment:
Yup, it looks like it will be even faster without the if-branch in the loop. I'm not sure if the compiler is smart enough to automatically/implicitly compile it that way or in similar ways (so that we don't have to optimize manually/explicitly).